hazelcast/src/test/java/com/hazelcast/nio/serialization/SerializationTest.java

```diff
@@ -500,4 +509,146 @@ public void readData(ObjectDataInput in) throws IOException {
+        ClassLoader oldContextLoader = Thread.currentThread().getContextClassLoader();
+        try {
+            ClassLoader current = getClass().getClassLoader();
+            DynamicProxyTestClassLoader cl = new DynamicProxyTestClassLoader(current);
+            Thread.currentThread().setContextClassLoader(cl);
+            SerializationService ss = new DefaultSerializationServiceBuilder().setClassLoader(cl).build();
+            IObjectA oa = (IObjectA) Proxy.newProxyInstance(current, new Class[] { IObjectA.class }, DummyInvocationHandler.INSTANCE);
+            Data data = ss.toData(oa);
+            Object o = ss.toObject(data);
+            Assert.assertSame("context classloader is not used", cl, o.getClass().getClassLoader());
+            try {
+                IObjectA.class.cast(o);
+                Assert.fail("the serialized object should not be castable");
+            } catch(ClassCastException e) {
+                // expected
+            }
+        } finally {
+            Thread.currentThread().setContextClassLoader(oldContextLoader);
+        }
+    }
+
+    @Test
+    public void testNonPublicDynamicProxySerialization_withClassLoaderMess() {
+        ClassLoader current = getClass().getClassLoader();
+        DynamicProxyTestClassLoader cl1 = new DynamicProxyTestClassLoader(current, IPrivateObjectC.class.getName());
+        DynamicProxyTestClassLoader cl2 = new DynamicProxyTestClassLoader(cl1, IPrivateObjectD.class.getName());
+        SerializationService ss = new DefaultSerializationServiceBuilder().setClassLoader(cl2).build();
+        Object ocd = Proxy.newProxyInstance(current, new Class[] { IPrivateObjectC.class, IPrivateObjectD.class }, DummyInvocationHandler.INSTANCE);
+        Data data = ss.toData(ocd);
+        try {
+            ss.toObject(data);
+            Assert.fail("the object should not be deserializable");
+        } catch (IllegalAccessError e) {
+            // expected
+        }
+    }
+
+    private static final class DynamicProxyTestClassLoader extends ClassLoader {
+
+        private static final Set<String> WELL_KNOWN_TEST_CLASSES = new HashSet<String>(Arrays.asList(IObjectA.class.getName(), IObjectB.class.getName(), IPrivateObjectC.class.getName(), IPrivateObjectD.class.getName()));
+
+        private final Set<String> wellKnownClasses = new HashSet<String>();
+
+        private DynamicProxyTestClassLoader(ClassLoader parent, String... classesToLoad) {
+            super(parent);
+            if (classesToLoad.length == 0) {
+                wellKnownClasses.addAll(WELL_KNOWN_TEST_CLASSES);
+            } else {
+                wellKnownClasses.addAll(Arrays.asList(classesToLoad));
+            }
+        }
+
+        protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+            if (!WELL_KNOWN_TEST_CLASSES.contains(name)) {
+                return super.loadClass(name, resolve);
+            }
+            synchronized (getClassLoadingLock(name)) {
```

Comment:
`getClassLoadingLock` is available since Java 7, can you please amend this to work on Java 6?